libbeat/autodiscover/autodiscover.go

```diff
@@ -0,0 +1,243 @@
+			return nil, err
+		}
+		logp.Debug(debugK, "Configured autodiscover provider: %s", provider)
+		providers = append(providers, provider)
+	}
+
+	return &Autodiscover{
+		bus:       bus,
+		adapter:   adapter,
+		runners:   cfgfile.NewRegistry(),
+		providers: providers,
+		meta:      make(map[uint64]*common.MapStrPointer),
+	}, nil
+}
+
+// Start autodiscovery process
+func (a *Autodiscover) Start() {
+	if a == nil {
+		return
+	}
+
+	logp.Info("Starting autodiscover manager")
+	a.startListener = a.bus.Subscribe(a.adapter.StartFilter()...)
+	a.stopListener = a.bus.Subscribe(a.adapter.StopFilter()...)
+
+	for _, provider := range a.providers {
+		provider.Start()
+	}
+
+	go a.startWorker()
+	go a.stopWorker()
+}
+
+func (a *Autodiscover) startWorker() {
+	for event := range a.startListener.Events() {
+		// This will happen on Stop:
+		if event == nil {
+			return
+		}
+
+		configs, err := a.adapter.CreateConfig(event)
+		if err != nil {
+			logp.Debug(debugK, "Could not generate config from event %v: %v", event, err)
+			continue
+		}
+		logp.Debug(debugK, "Got a start event: %v, generated configs: %+v", event, configs)
+
+		meta := getMeta(event)
+		for _, config := range configs {
+			hash, err := hashstructure.Hash(config, nil)
+			if err != nil {
+				logp.Debug(debugK, "Could not hash config %v: %v", config, err)
+				continue
+			}
+
+			if a.runners.Has(hash) {
+				logp.Debug(debugK, "Config %v is already running", config)
+				continue
+			}
+
+			err = a.adapter.CheckConfig(config)
+			if err != nil {
+				logp.Debug(debugK, "Check failed for config %v: %v, won't start runner", config, err)
+				continue
+			}
+
+			runner, err := a.adapter.Create(config, a.storeMeta(hash, meta))
+			if err != nil {
+				logp.Debug(debugK, "Failed to create runner with config %v: %v", config, err)
+				continue
+			}
+
+			logp.Info("Autodiscover starting runner: %s", runner)
+			a.runners.Add(hash, runner)
+			runner.Start()
+		}
+	}
+}
+
+func (a *Autodiscover) stopWorker() {
+	for event := range a.stopListener.Events() {
+		// This will happen on Stop:
+		if event == nil {
+			return
+		}
+
+		configs, err := a.adapter.CreateConfig(event)
+		if err != nil {
+			logp.Debug(debugK, "Could not generate config from event %v: %v", event, err)
+			continue
+		}
+		logp.Debug(debugK, "Got a stop event: %v, generated configs: %+v", event, configs)
+
+		for _, config := range configs {
+			hash, err := hashstructure.Hash(config, nil)
+			if err != nil {
+				logp.Debug(debugK, "Could not hash config %v: %v", config, err)
+				continue
+			}
+
+			if !a.runners.Has(hash) {
+				logp.Debug(debugK, "Config %v is not running", config)
+				continue
+			}
+
+			if runner := a.runners.Get(hash); runner != nil {
+				logp.Info("Autodiscover stopping runner: %s", runner)
+				runner.Stop()
+				a.runners.Remove(hash)
+			} else {
+				logp.Debug(debugK, "Runner not found for stopping: %s", hash)
+			}
+		}
+	}
+}
+
+func getMeta(event bus.Event) *common.MapStr {
```

Comment:
Should this be a method on `bus.Event`?

Reply:
bus.Event is quite generic and can be used in other parts, `meta` field is just a convention for autodiscover. I would prefer to keep bus slim until we see how it gets used